Token Request Builder

aguevara23 opened the pull request:
## Why do we need this change?
Token builder API expects a POST request with tokens in the post body. 


## Implementation Details
- This change adds some helper methods and classes for generating post body payload that Sorcerer expects.
- When a token class is instantiated it will run validations for passed in arguments. 
- Karma config was updated to include nested files inside the `test` directory.

TODO:
- update documentation, update README


#### Dependencies (if any)
none

:house: [ch55669](https://app.clubhouse.io/movableink/story/55669)

MansurTsutiev committed:
feat: add token types and RequestBuilder

- update karma config to include sub-directories

Co-authored-by: Alex Guevara <aguevara@movableink.com>

MansurTsutiev committed:
feat: update RequestBuilder tests with new token types

- get rid of the `.` period when building errors

Co-authored-by: Alex Guevara <aguevara@movableink.com>

shyshy reviewed (changes requested):
Looks good directionally.

There's a couple required pieces missing from this PR that will need to be done before this can get merged.

We will need to make sure this introduces the correct changes to make a clean release.

1. Transfer the Notion documentation on tokens into this repo. Somewhere in a Wiki maybe
2. Update the README to point to this documentation
3. Update the changelog in the README
4. Update `package.json` with a version bump (i believe this is a backwards compatible minor version update)

There might be other things that I'm forgetting from the release cycle. Something to just think  about and review.

aguevara23 commented on src/token-builder/request-builder.js:
```diff
@@ -0,0 +1,25 @@
+    });
+
+    if (errors.length) {
+      throw new Error(`Errors found while parsing tokens:\n${errors.join('\n')}`);
```
How does something like this sound?
```
Request was made with invalid tokens. See validation errors below:
token 1: Missing properties for replace token...,
token2: Missing properties for secret token...
```

shyshy replied:
That sounds closer! I think `Request was built and not executed due to invalid tokens...` is closer too.

aguevara23 commented on src/token-builder/types.js:
```diff
@@ -0,0 +1,192 @@
+    this.validateOptions();
+  }
+
+  toJSON() {
```
We used to have `toJSON` on `TokenBase` but removed it because a user wouldn't actually be instantiating a base token. Though should we still have that method on the base class?

Note: for the `toJSON` method on each token subclass, 
- `replace` and `replaceLarge` return  the same properties
- `hmac` and `sha1` return the same propertie
- `secret` token is unique

aguevara23 committed:
default hmacOptions to empty object

aguevara23 committed:
add toJSON method to TokenBase

aguevara23 committed:
update error message in RequestBuilder

aguevara23 committed:
update how Replace and ReplaceLarge 'value' property is validated

aguevara23 committed:
update version in package.jsonn

aguevara23 commented on src/token-builder/types.js:
```diff
@@ -0,0 +1,192 @@
+  validateOptions() {
+    const missingProps = [];
+    this.requiredProperties.forEach((prop) => {
+      if (!this[prop]) {
```
Since `value` is the only property where an empty string is valid, I opted to keep this logic in place for `TokenBase`, but update how the validations are done for the `Replace` and `ReplaceLarge` tokens.

aguevara23 committed:
update readme

aguevara23 committed:
remove references to tokenName for hmac tokens

MansurTsutiev committed:
refactor: perform a single check for nullish values

aguevara23 committed:
add documentation for token-builder

shyshy commented on docs/token-builder.md:
```diff
@@ -0,0 +1,288 @@
+
+Currently supported tokens are:
+
+- ReplaceToken
```
suggestion (non-blocking): [linking to the same section](https://stackoverflow.com/a/16426829) can be a good way to shortcut to the relevant part of the doc.

shyshy commented on docs/token-builder.md:
```diff
@@ -0,0 +1,288 @@
+
+Replaces HMAC token with an HMAC signature generated from the token options
+
+O**ptions:**
```
fix: small typo here

shyshy replied:
suggestion: on a similar note to my previous comments about "documentation best practices", I think we should be more explicit about the "structure" of `options`, vs. `hmacOptions`.

I think you should consider defining a top level `params` section that describes the options that are native to each token, things like `cacheOverride`, `skipCache`, `name`, etc., with explanations of each and what defaults they get.

I think only _then_, is it less confusing to describe what `options` is _within_ these `params`. Otherwise it just sounds like "options within options".

shyshy commented on docs/token-builder.md:
```diff
@@ -0,0 +1,288 @@
+tokenModel.toJSON() // returns:
+// {
+//   name: 'FavoriteBand',
+//   type: 'replace',
```
fix: this should result in `replaceLarge` or whichever is the one that `sorcerer` considers is "replace with large value"

shyshy commented on docs/token-builder.md:
````diff
@@ -0,0 +1,288 @@
+```jsx
+const options = {
+  name: 'FavoriteBand',
+  cacheOverride: 'Movable Band',
````
suggestion: When it comes to documentation, IMO as a best practice, we should start off with the "simplest" examples. In this case, we should strip out `cacheOverride` and `skipCache`. We should let them assume the defaults of what the classes were meant to do, and provide more advanced examples further down in their respective explanations for `cacheOverride` and `skipCache`.

The downside in providing them initially is that users think that this is the preferred usage. We should encourage the default usage and only provide how to non-default options as exceptions (with guidance on how the exception works).

shyshy commented on docs/token-builder.md:
```diff
@@ -0,0 +1,288 @@
+//   type: 'secret',
+//   path: 'watson',
+//   skipCache: true,
+//   cacheOverride: 'xyz',
```
suggestion: can we exclude `cacheOverride` from this? see above explanation.

shyshy commented on docs/token-builder.md:
```diff
@@ -0,0 +1,288 @@
+
+    Once the signature is generated it needs to be encoded
+
+    The following encodings are supported:`hex`, `base64`,`base64url` ,`base64percent`
```
fix: I think the spacing on this sentence is a bit off between the `:` and the `,`.

shyshy commented on docs/token-builder.md:
```diff
@@ -0,0 +1,288 @@
+
+Each token has a couple of properties for handling how the token is included when generating a cache key:
+
+- **skipCache** - if set to `true`, then the token will not be considered when caching requests
```
suggestion: This is where I would show examples of how to provide these options when constructing a token.

shyshy reviewed (changes requested):
Docs look good. Added some fixes, nits, suggestions.

aguevara23 committed:
fix: typos, inconsistent styling

aguevara23 committed:
chore: update token examples in docs, add examples to caching section

aguevara23 committed:
chore: add params section to documentation

aguevara23 committed:
chore: add char limit validation for cacheOverride

Commit:
Merge pull request #49 from movableink/ag/cache-override-validation

 chore: set char limit on cacheOverride

aguevara23 committed:
update check for valid cacheOverride

aguevara23 committed:
update documentation

- update README table of contents changelog to link
to new version
- update Token Builder docs with toc, new examples

- Co-authored-by: Mansur Tsutiev <mtsutiev@movableink.com>

aguevara23 committed:
export TokenBuilder classes from index.js

Co-authored-by: Mansur Tsutiev <mtsutiev@movableink.com>

aguevara23 committed:
refactor: perform a single check for nullish values for ReplaceLarge token

MansurTsutiev committed:
refactor: use null instead of undefined

it's more common to use null instead of undefined
and it's less characters :)
